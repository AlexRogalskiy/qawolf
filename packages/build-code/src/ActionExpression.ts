export class ActionExpression {
  code() {
    return "";
  }

  updatableCode() {
    return this.code();
  }
}
