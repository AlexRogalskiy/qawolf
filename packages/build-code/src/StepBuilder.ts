import { buildSteps } from "@qawolf/build-workflow";
import { logger } from "@qawolf/logger";
import { ElementEvent, Step } from "@qawolf/types";
import { sortBy } from "lodash";

type ConstructorOptions = {
  isTest?: boolean;
  startIndex?: number;
};

export class StepBuilder {
  /**
   * Build steps as we receive events.
   */
  private _events: ElementEvent[] = [];
  private _finalized: boolean = false;
  private _steps: Step[] = [];

  // used for code with existing steps
  private _startIndex: number;

  public constructor({ startIndex }: ConstructorOptions = {}) {
    this._startIndex = startIndex || 0;
  }

  private _buildSteps({ canChange }: { canChange: boolean }) {
    const steps = buildSteps({
      canChange,
      events: this.getEvents(),
      startIndex: this._startIndex
    });

    const newSteps = steps.slice(this._steps.length);
    if (!newSteps.length) return;

    logger.debug(`StepTracker: built new steps ${JSON.stringify(newSteps)}`);
    this._steps.push(...newSteps);
  }

  public finalize() {
    this._finalized = true;

    this._buildSteps({
      // there are no new events coming in
      // so we know the steps that canChange will not
      canChange: true
    });
  }

  public getEvents() {
    return sortBy(this._events, e => e.time);
  }

  public pushEvent(event: ElementEvent) {
    if (this._finalized) return;

    this._events.push(event);
    this._buildSteps({ canChange: false });
  }

  public steps() {
    return this._steps;
  }
}
