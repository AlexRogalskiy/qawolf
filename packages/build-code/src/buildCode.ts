import { Workflow } from "@qawolf/types";
import { readFileSync } from "fs-extra";
import { compile } from "handlebars";
import { camelCase } from "lodash";
import { resolve } from "path";
import { formatLaunch } from "./formatLaunch";
import { formatStep } from "./formatStep";

type BuildCodeOptions = {
  test?: boolean;
  workflow: Workflow;
};

const scriptTemplate = compile(
  readFileSync(resolve(__dirname, "../static/script.hbs"), "utf8")
);

const testTemplate = compile(
  readFileSync(resolve(__dirname, "../static/test.hbs"), "utf8")
);

export const buildCode = (options: BuildCodeOptions) => {
  const { workflow } = options;

  const template = options.test ? testTemplate : scriptTemplate;

  const code = template({
    launch: formatLaunch(workflow.url, workflow.device),
    name: camelCase(workflow.name),
    steps: workflow.steps.map((step, i) => {
      const previousStep = i > 0 ? workflow.steps[i - 1] : null;

      return formatStep(step, previousStep);
    }),
    url: workflow.url
  });

  return code;
};
