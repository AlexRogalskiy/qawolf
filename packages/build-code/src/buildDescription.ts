import { Step } from "@qawolf/types";
import { describeDoc } from "@qawolf/web";
import { isNil } from "lodash";

export const buildDescription = (step: Step): string => {
  if (step.action === "scroll") {
    return `scroll`;
  }

  const description = describeDoc(step.html);

  const target = step.html.node;

  // link/input
  const tagName = `${target.name === "a" ? " link" : ` ${target.name}` || ""}`;

  if (step.action === "click") {
    return `click${description}${tagName}`;
  }

  if (step.action === "select") {
    return `select${description}`;
  }

  if (step.action === "type") {
    if (step.replace && (step.value === "" || isNil(step.value))) {
      return `clear${description}${tagName}`;
    }

    const value = step.value as string;
    if (value && value.indexOf("↓Enter") === 0) {
      return `Enter`;
    }

    if (value && value.indexOf("↓Tab") === 0) {
      return `Tab`;
    }

    return `type into${description}${tagName}`;
  }

  throw new Error(`Invalid step action ${step.action}`);
};
