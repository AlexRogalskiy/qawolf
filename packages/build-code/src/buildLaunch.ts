export const buildLaunch = (url: string, device?: string): string => {
  if (device) {
    return `await launch({ device: "${device}", url: "${url}" });`;
  }

  return `await launch({ url: "${url}" });`;
};
