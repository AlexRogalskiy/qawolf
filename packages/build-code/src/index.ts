export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
