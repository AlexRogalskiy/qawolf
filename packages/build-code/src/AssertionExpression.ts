import { ActionExpression } from "./ActionExpression";

type ConstructorOptions = {
  action: ActionExpression;
  description: string;
};

export class AssertionExpression {
  private _action: ActionExpression;
  private _description: string;

  constructor({ action, description }: ConstructorOptions) {
    this._action = action;
    this._description = description;
  }

  private _formatIt() {
    const description = this._description;
    if (!description.includes(`"`)) return `"can ${description}"`;
    if (!description.includes("'")) return `'can ${description}'`;
    return JSON.stringify("can " + description);
  }

  code() {
    const code = `\nit(${this._formatIt()}, async () => {\n  ${this._action.code()}\n  });\n`;
    return code;
  }

  updatableCode() {
    return this._action.updatableCode();
  }
}
