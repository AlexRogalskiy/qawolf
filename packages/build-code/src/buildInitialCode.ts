import { camelCase } from "lodash";
import { buildLaunch } from "./buildLaunch";

export type InitialCodeOptions = {
  device?: string;
  isTest?: boolean;
  name: string;
  patchHandle: string;
  url: string;
};

type TemplateOptions = {
  launch: string;
  name: string;
  patchHandle: string;
};

const buildInitialScript = ({ name, launch, patchHandle }: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

const ${name} = async () => {
  const browser = ${launch}
  ${patchHandle}
  await browser.close();
};

${name}();`;

  return code;
};

const buildInitialTest = ({ name, launch, patchHandle }: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

describe('${name}', () => {
  let browser;

  beforeAll(async () => {
    browser = ${launch}
  });

  afterAll(() => browser.close());
  ${patchHandle}
});`;

  return code;
};

export const buildInitialCode = (options: InitialCodeOptions) => {
  const templateOptions = {
    launch: buildLaunch(options.url, options.device),
    name: camelCase(options.name),
    patchHandle: options.patchHandle
  };

  if (options.isTest) {
    return buildInitialTest(templateOptions);
  }

  return buildInitialScript(templateOptions);
};
