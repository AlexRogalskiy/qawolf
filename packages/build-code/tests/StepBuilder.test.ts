import { loadEvents } from "@qawolf/fixtures";
import { ElementEvent } from "@qawolf/types";
import { StepBuilder } from "../src/StepBuilder";

describe("StepBuilder", () => {
  let events: ElementEvent[];

  beforeAll(async () => {
    events = await loadEvents("scroll_login");
  });

  it("builds steps for events", () => {
    // should start steps at 10
    const builder = new StepBuilder({ startIndex: 10 });
    expect(builder.steps()).toEqual([]);

    const splitIndex = events.length / 2;

    events.slice(0, splitIndex).forEach(event => builder.pushEvent(event));
    const stepsOne = builder.steps().slice();
    expect(stepsOne).toMatchSnapshot();
    expect(stepsOne).toHaveLength(2);

    events.slice(splitIndex).forEach(event => builder.pushEvent(event));
    const stepsTwo = builder.steps().slice(stepsOne.length);
    expect(stepsTwo).toMatchSnapshot();
    expect(stepsTwo).toHaveLength(6);
  });
});
