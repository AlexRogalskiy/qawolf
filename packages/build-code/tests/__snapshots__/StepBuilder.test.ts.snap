// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepBuilder builds steps for events 1`] = `
Array [
  Object {
    "action": "scroll",
    "canChange": false,
    "html": Object {
      "ancestors": Array [],
      "node": Object {
        "attrs": Object {},
        "children": Array [],
        "name": "html",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 10,
    "page": 0,
    "value": Object {
      "x": 0,
      "y": 378,
    },
  },
  Object {
    "action": "click",
    "canChange": false,
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {},
          "children": Array [],
          "name": "li",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {},
          "children": Array [],
          "name": "ul",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "href": "/login",
          "innertext": "Form Authentication",
        },
        "children": Array [
          Object {
            "content": "Form Authentication",
            "type": "text",
          },
        ],
        "name": "a",
        "type": "tag",
        "voidElement": false,
      },
    },
    "index": 11,
    "page": 0,
  },
]
`;

exports[`StepBuilder builds steps for events 2`] = `
Array [
  Object {
    "action": "click",
    "canChange": false,
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "large-6 small-12 columns",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "row",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "username",
          "labels": "Username",
          "name": "username",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 12,
    "page": 0,
  },
  Object {
    "action": "type",
    "canChange": false,
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "large-6 small-12 columns",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "row",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "username",
          "labels": "Username",
          "name": "username",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 13,
    "page": 0,
    "value": "tomsmith",
  },
  Object {
    "action": "type",
    "canChange": false,
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "large-6 small-12 columns",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "row",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "username",
          "labels": "Username",
          "name": "username",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 14,
    "page": 0,
    "value": "↓Tab↑Tab",
  },
  Object {
    "action": "type",
    "canChange": false,
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "large-6 small-12 columns",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "row",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "password",
          "labels": "Password",
          "name": "password",
          "type": "password",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 15,
    "page": 0,
    "value": "SuperSecretPassword!",
  },
  Object {
    "action": "type",
    "canChange": false,
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "large-6 small-12 columns",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "row",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "id": "password",
          "labels": "Password",
          "name": "password",
          "type": "password",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
    },
    "index": 16,
    "page": 0,
    "value": "↓Enter↑Enter",
  },
  Object {
    "action": "click",
    "canChange": false,
    "html": Object {
      "ancestors": Array [
        Object {
          "attrs": Object {
            "class": "example",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
        Object {
          "attrs": Object {
            "class": "large-12 columns",
            "id": "content",
          },
          "children": Array [],
          "name": "div",
          "type": "tag",
          "voidElement": false,
        },
      ],
      "node": Object {
        "attrs": Object {
          "class": "button secondary radius",
          "href": "/logout",
          "innertext": "Logout",
        },
        "children": Array [
          Object {
            "attrs": Object {
              "class": "icon-2x icon-signout",
            },
            "children": Array [
              Object {
                "content": " Logout",
                "type": "text",
              },
            ],
            "name": "i",
            "type": "tag",
            "voidElement": false,
          },
        ],
        "name": "a",
        "type": "tag",
        "voidElement": false,
      },
    },
    "index": 17,
    "page": 0,
  },
]
`;
